### ISSUE TITLE
_Alignas in typedef is no longer usable for VST's purposes

### ISSUE THREAD

**Issue by andrew-appel** [AbsInt/CompCert#256]
The resolution of issue #197 was that CompCert _will_ support _Alignas in a typedef, even though that's not part of the C standard, for reasons explained in that issue.  But now it appears that the current master branch does not support this anymore.    This is a bug.

@txyyss reports,   Recently I tried to use the clightgen built from the master branch of CompCert to recompile our source code that works in CompCert 3.3.  It seems that it does not support _Alignas in typedef now.

It reports `"error: alignment specified for typedef 'value'"` for the following definition:

```
typedef void  * value
#ifdef COMPCERT
  _Alignas (void *)
#endif
  ;
```

If I change the definition to 

```
typedef 
#ifdef COMPCERT
  _Alignas (void *)
#endif
  void  * value;
```

there is no complaint. But the generated file only has (tptr tvoid) instead of (talignas 2%N (tptr tvoid)).

**Comment by bschommer** [comment]
Unfortunately we need to be more compliant with the C standard, so ```_Alignas``` is no longer supported in non standard locations, for example typedef.

**Comment by xavierleroy** [comment]
However, in an attempt to balance conformance with usability and to imitate gcc and clang, we're trying to support `__attribute((aligned(N)))` in all contexts, including typedefs.  So, the following should keep working:
```
typedef void * value __attribute((aligned(N)));
```
`N` is a compile-time integer constant; you can use
```
__attribute((aligned(_Alignof(ty))))
```
as a replacement for `_Alignas(ty)`.

We're still tweaking the handling of `_Alignas` and the `aligned` attribute for maximal C11 conformance and maximal gcc/clang compatibility.  There might be some other changes in the near future.

**Comment by txyyss** [comment]
Yes, it works. Thank you very much!

Shengyi Wang

Cell: +65-83509639


Xavier Leroy <notifications@github.com> 于2018年9月6日周四 下午9:50写道：

**Comment by andrew-appel** [comment]
I'm sorry to have to reopen this issue, but it is still broken.

Consider this:
```
typedef void *int_or_ptr  __attribute((aligned(_Alignof(void*))));

int_or_ptr f (int_or_ptr x) {int_or_ptr a; ... }
```
When translated by clightgen in CompCert 3.4, the return type of the function is simply "pointer to void", without the attribute.   The parameter `x` and local variable `a` do have the attribute.

This is very awkward indeed, because (in a proof about the program), the type that the user gave for the function (int_or_ptr) is not the type in the abstract syntax tree.

Writing 
`int_or_ptr __attribute((aligned(_Alignof(void*))))  f (int_or_ptr x) {int_or_ptr a; ... }`
doesn't work either; the return type does not have the attribute.

Up through CompCert 3.3, using `typedef void  * value  _Alignas (void *) ;`
this worked:  the return type, the parameter, and the local variable all had the attribute.

It is an important feature of VST to support int-or-pointer values.  They are not part of the C11 standard, but (in effect) gcc and clang support them, and they are essential to the implementation of many garbage-collected languages, including Ocaml (and hence,  Coq, and hence, the foundations of CompCert itself).    The agreed resolution of issue #197 was, CompCert will support an attribute syntax of typedef that allows to distinguish the int_or_ptr type from void*, (even though they are semantically equivalent).  CompCert 3.4 fails to fully support that.  Based on that agreement, many weeks of work have been invested in VST and a garbage-collector proof that makes use of this feature.

Side note: At present, the correctness proof of CompCert does not support testing the low-order bit to distinguish an odd integer from an aligned pointer, so the user of int-or-pointer values must add an axiom.   There is some hope that the correctness proof can support them in the future (by more careful reasoning about memory injections), and then it will be possible to remove the axiom.  This is really a separate issue from the failure to support an int_or_ptr type in CompCert.

So, in summary:  Is there a way to fix the function-return type declared using a typedef-with-attribute, so the attribute is not lost?

**Comment by andrew-appel** [comment]
By the way, not only the return-type of the function has lost the attribute, but in any function call to this function, the temporary variable inserted in the Clight program is also missing the attribute.

**Comment by xavierleroy** [comment]
There seems to be several misunderstandings here.  Let me clarify them, in this comment and in the next one.

First comment: your use of alignment specifiers makes no sense.  When you write (in CompCert 3.3)
```
typedef void * value _Alignas (void *) ;
```
you're asking for a `void *` type that is aligned like a `void *` type.  This is completely equivalent to
```
typedef void * value;
```
So, why bother with alignment modifiers at all?

Likewise, I understand your desire to mix integer values with pointer values in a common type.  Just use `intptr_t` or `uintptr_t`: these are integer types that are guaranteed to have the same size and alignment as pointer types.  You can `typedef void * value` and cast to/from `uintptr_t` when you'd like to do integer operations, or `typedef uintptr_t value` and cast to/from `T *` when you'd like to use as an address.  There is no need whatsoever to play with alignment modifiers.

**Comment by xavierleroy** [comment]
Second comment: you have observed that `_Alignas` and `__attribute((aligned(N)))` apply to names in CompCert 3.4 while they apply to types in CompCert 3.3.  I stated that earlier but it seems that more explanations are in order.

The C syntax for type modifiers/qualifiers is somewhat ambiguous.  Consider:
```
const int * f(void);
_Noreturn int * g(void);
```
`const` qualifies (applies to) `int`, so `f` has type "function returning a pointer to a const int".
`_Noreturn` applies to a function, so `g` has type "Noreturn function returning a pointer to an int".

In the examples above, which are standard C11, each keyword has an "attachment point": `const` goes to types, `_Noreturn` goes to functions. 

In contrast, GCC attributes all share the same keyword `__attribute`; it's the name of the attribute that determines where they apply and what part of the AST they attach to.  A type attribute attaches to types the way `const` does.  A name attribute attaches to (the type of) the variable or function being declared.  Etc.

Up to CompCert 3.3, `_Alignas` and `__attribute((aligned(N)))` were treated as type attributes, attaching to types like `const` does.  That was a bad decision because it is incompatible with C11 (for `_Alignas`) and with GCC (for `__attribute((aligned(N)))`.  E.g.
```
_Alignas(16) int * p;
```
was "p is a pointer to 16-aligned integers" in CompCert 3.3, but is "p is 16-aligned and is a pointer to integers" in C11 and in CompCert 3.4.

The one remaining GCC oddity is the "aligned" attribute in a typedef.  This is best illustrated by examples:
```
typedef __attribute(aligned((16)) int intal16;
intal16 * p;
__attribute(aligned((16)) int * q;
```
In CompCert, both `p` and `q` are 16-aligned pointers to integers.
In GCC, `q` is a 16-aligned pointer, but `p` is a naturally-aligned pointer to 16-aligned integers.
I think GCC's interpretation makes no sense: a typedef should always behave like its textual expansion at point of use.

So, I'm sorry if I'm being stubborn about it, but I think CompCert 3.3's behavior was incorrect and CompCert 3.4's is the correct one.  To go further we might add a warning or even an error for
```
typedef __attribute(aligned((16)) int intal16;
```
or any typedef that contains non-type-related attributes, because it's just confusing (name-related attributes, for example, should be written in the declaration of the name, not hidden in a typedef) and a source of incompatibility with GCC.

**Comment by andrew-appel** [comment]
To answer @xavierleroy's question, "you're asking for a void * type that is aligned like a void * type", these are semantically the same, so "why bother with alignment modifiers at all?"

This is explained clearly in the first post of issue #197; and in the third post of #197 @xavierleroy wrote, "I think this is an innocuous and potentially useful extension to C11, so why not support it? and even mention it in the manual, indeed? ... The usage outlined by @andrew-appel is kind of the poor man's custom attribute in Clight."

If, in August of 2017, the answer had been, "it is inappropriate to support this," then we would not have invested weeks of effort in VST to make use of this feature.  But the answer was, "Added a mention to user's manual, will be in release 3.1."

**Comment by bschommer** [comment]
Unfortunately the conclusion in #197 

> Again I think this is an innocuous and potentially useful extension to C11,  so why not support it? and even mention it in the manual, indeed?

was wrong and that is why the behavior change. First of all in the case of ``_Alignas`` we need to be compliant with what the C 11 standard allows as usage and for the `__attribute(alined)`  it has shown that the old behavior of CompCert was not the one used by gcc and the result was that CompCert was slightly incompatible with gcc.

**Comment by xavierleroy** [comment]
I miremembered the discussion in #197, which does give the background I was missing.

The conclusion about "the poor man's custom attribute in Clight" remains true at the AST level; it's just that there is no (longer) any concrete syntax to put that attribute where you want it.  

The conclusion about "innocuous extension to C11" was very wrong, as @bschommer explains.

Right now I have other things to worry about, so I suggest VST stays with CompCert 3.3 for a while.

In the medium term I'm sure we can find a solution, preferably one that is completely contained in clightgen and calls for no modifications to the main compiler base.  Our hands are increasingly tied by qualification requirements and other reasons.




More generally, it's just that GCC's attribute syntax is awful, as it gives no way to control exactly where the attribute attaches in the AST.

**Comment by xavierleroy** [comment]
I changed the title of this issue to something more descriptive.  I agree to reopen it if you agree to never use the word "broken" again.

**Comment by xavierleroy** [comment]
CompCert's Git master contains a series of commits (0616171, d7dfc21, 4fb8df8) whose purpose is to make handling of attributes closer to what GCC and Clang do.  In particular, the `aligned` attribute now attaches to `typedef` the GCC way.  As a consequence, the problematic example
```
typedef void *int_or_ptr  __attribute((aligned(_Alignof(void*))));

int_or_ptr f (int_or_ptr x) {int_or_ptr a; return x; }
```
is now translated by clightgen to
```
Definition f_f := {|
  fn_return := (talignas 3%N (tptr tvoid));
  fn_callconv := cc_default;
  fn_params := ((_x, (talignas 3%N (tptr tvoid))) :: nil);
  fn_vars := nil;
  fn_temps := ((_a, (talignas 3%N (tptr tvoid))) :: nil);
  fn_body := (Sreturn (Some (Etempvar _x (talignas 3%N (tptr tvoid)))))
|}.
```
which I believe is the expected behavior for VST's purposes.

Please test and confirm that this is OK or explain me why this still doesn't do what VST needs.  I'm planning a public release 3.5 of CompCert later this week that will include this revised handling of attributes.

**Comment by andrew-appel** [comment]
I don't have time this week to test it.  However, your example f_f looks like just the desired behavior, so I anticipate with pleasure the release of CompCert 3.5.

**Comment by txyyss** [comment]
Anshuman and me both tested the latest commit of CompCert with the latest
commit of VST. It works perfectly with our CertiGC proofs.

The .v file generated by clightgen is exactly the same as before, except
the version number.

Thank you very much,

Shengyi Wang

Cell: +65-83509639


Andrew Appel <notifications@github.com> 于2019年2月26日周二 上午3:09写道：

**Comment by xavierleroy** [comment]
Thanks for the testing.  Let's say this is fixed in release 3.5.
